### PR TITLE
IIFU 230810 (Rebase)

### PR DESCRIPTION
**Pull request by jayfurmanek**
#300 Rebased and cleaned up

closes #300

### CODE REVIEW

File: lib/Dialect/TritonGPU/Transforms/AccelerateMatmul.cpp [hunk 1]
```diff
@@ -279,36 +304,28 @@ class BlockedToMMA : public mlir::RewritePattern {
...
+    Type minType = IntegerType::get(ctx, minBitwidth);
+    // convert A operand
     auto newAEncoding = triton::gpu::DotOperandEncodingAttr::get(
         oldAType.getContext(), 0, newRetType.getEncoding(),
-        oldAType.getElementType());
-    auto newBEncoding = triton::gpu::DotOperandEncodingAttr::get(
-        oldBType.getContext(), 1, newRetType.getEncoding(),
-        oldBType.getElementType());
-
+        minBitwidth > 0 ? minType : oldAType.getElementType());
```
Comment (zhanglx13) [inline lib/Dialect/TritonGPU/Transforms/AccelerateMatmul.cpp:317]: With this change getting into our fork, we will be using the [other constructor](https://github.com/ROCmSoftwarePlatform/triton/blob/9cdf3a58c396a8265a788d4b5a5672d9b0c0c22a/include/triton/Dialect/TritonGPU/IR/TritonGPUAttrDefs.td#L642C16-L642C39) for dotOperand, which sets `kWidth` explicitly as 32 / elemBitWidth.
@binarman Can you double check if this is the case? If so, we'll probably need some fixes there.

File: lib/Conversion/TritonGPUToLLVM/ScanOpToLLVM.cpp [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+// compute a flat id for the parallel dimensions.
+std::tuple<Value, Value, Value>
+ScanOpConversion::getDelinearizedIds(ConversionPatternRewriter &rewriter,
+                                     ScanLoweringHelper &helper) const {
+  auto loc = helper.getLoc();
+  unsigned axis = helper.getAxis();
+  auto srcEncoding = helper.getEncoding();
+
+  Value threadId = getThreadId(rewriter, loc);
+  Value warpSize = i32_val(32);
```
Comment (zhanglx13) [inline lib/Conversion/TritonGPUToLLVM/ScanOpToLLVM.cpp:246]: So we just grabbed everything about scanOp from upstream? At least the warpSize needs to be adjusted. But I'm not sure if we also need some algorithmic level modifications for scan to make it work better on AMD gpus.

File: lib/Dialect/TritonGPU/Transforms/AccelerateMatmul.cpp (hunk 1 above)
Comment (binarman) [reply lib/Dialect/TritonGPU/Transforms/AccelerateMatmul.cpp:317]: This function work only for MMA layout, so it should not affect MFMA layout.
This change is safe.

Comment (binarman) [reply lib/Dialect/TritonGPU/Transforms/AccelerateMatmul.cpp:317]: In current triton-mlir - nowhere =)
I've added it in https://github.com/ROCmSoftwarePlatform/triton/pull/295/files#diff-980bbab8870c6a2f27dbee8696cea2a5d4db63de32bf63f6584ac1fbd3409d9fR209, it is not merged yet

Comment (jayfurmanek) [reply lib/Dialect/TritonGPU/Transforms/AccelerateMatmul.cpp:317]: yes, we can revisit in #295 then once this one is merged.

File: lib/Conversion/TritonGPUToLLVM/ScanOpToLLVM.cpp (hunk 2 above)
Comment (jayfurmanek) [reply lib/Conversion/TritonGPUToLLVM/ScanOpToLLVM.cpp:246]: Yup. This is a brand new file. I suspect it doesn't quite work yet. I'll open an issue to test/investigate/fix.

Comment (zhanglx13) [reply lib/Conversion/TritonGPUToLLVM/ScanOpToLLVM.cpp:246]: Here is the ticket for scanOp bringup: https://github.com/ROCmSoftwarePlatform/frameworks-internal/issues/4898

**Review by zhanglx13 — APPROVED** [review]

**Pull request merged**